packages/protocol/contracts/team/LibTokenGrant.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+library LibTokenGrant {
+    error INVALID_PARAMS();
+
+    function calcVestedAmount(
+        uint256 _grantAmount,
+        uint256 _vestDuration,
+        uint256 _time
+    )
+        internal
+        pure
+        returns (uint256)
+    {
+        if (_vestDuration == 0) return _grantAmount;
+        return _portion(_grantAmount, _time, _vestDuration);
+    }
+
+    function calcUnlockedAmount(
+        uint256 _grantAmount,
+        uint256 _vestDuration,
+        uint256 _unlockDuration,
```

Comment:
Is there some explanation of how the unlocking is supposed to work? Going on the tests, I don't think I understand.

Reply:
<img width="768" alt="Screenshot 2024-04-17 at 23 44 56" src="https://github.com/taikoxyz/taiko-mono/assets/99078276/0e5ce39a-2813-468d-a79e-8d8c140344fd">

Let me explain using the diagram above.

The green area represents the token grant total amount. The x-axis represents the time.

- ``a: the vesting duration
- `b`: the unlocking duration
- `a + b` then represents when all tokens will be vested AND unlocked.

The idea is that if some tokens are vested, these tokens' unlock period starts immediately at the vesting time:
- tokens vested at the first second (time 0) will have a unlocking period `[0, b]`;
- tokens vested at the last second (time a) will have a unlocking period `[a, a+b]`